src/include/miopen/miopen_internal.h

```diff
@@ -0,0 +1,114 @@
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef GUARD_MIOPEN_INTERNAL_H_
+#define GUARD_MIOPEN_INTERNAL_H_
+
+/* Put experimental APIs here. */
+/* If used, should be included after miopen.h. */
+
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wextern-c-compat"
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* Begin of Find Mode API */
+
+/*! @brief Version of the Find Mode API for convolutions.
+*
+* The Find Mode API is experimental and therefore subject to change.
+* This macro allows applications to adapt to the future.
+*/
+#define MIOPEN_API_VERSION_CONVOLUTION_FIND_MODE 1
+
+/*! @enum miopenConvolutionFindMode_t
+*
+* * Normal: This is the full Find mode call, which will benchmark all the solvers and return a list.
+*
+* * Fast: Checks the Find-db for an entry. If there is a hit, use that entry. If there is a miss,
+* utilize the Immediate mode fallback. If Start-up times are expected to be faster, but worse GPU
+* performance.
+*
+* * Hybrid: Checks the Find-db for an entry. If there is a hit, use that entry. If there is a miss,
+* use the existing Find machinery. Slower start-up times than Fast Find, but no GPU performance
+* drop.
+*
+* * Fast Hybrid: Checks the Find-db for an entry. If there is a hit, use that entry. If there is a
+* miss, uses the existing Find machinery with skipping slow-compiling kernels. Faster start-up times
+* than Hybrid Find, but GPU performance is a bit worse.
+*
+* * Dynamic Hybrid: This mode is similar to Fast Hybrid, but in case of Find-db miss skips all
+* non-dynamic kernels, thus saving compilation time. Versus Fast Hybrid, we expect similar start-up
+* times but better GPU performance.
+*/
+typedef enum {
+    miopenConvolutionFindModeNormal        = 1,                         /*!< Normal mode */
+    miopenConvolutionFindModeFast          = 2,                         /*!< Fast mode */
+    miopenConvolutionFindModeHybrid        = 3,                         /*!< Hybrid mode */
+    miopenConvolutionFindModeFastHybrid    = 4,                         /*!< Fast Hybrid mode */
+    miopenConvolutionFindModeDynamicHybrid = 5,                         /*!< Dynamic Hybrid mode */
+    miopenConvolutionFindModeDefault = miopenConvolutionFindModeHybrid, /*!< Default setting */
+} miopenConvolutionFindMode_t;
```

Comment:
Can we define this as an `int` and use constants instead?

Reply:
Hmm, I do not think we have constants in C. Or you mean something else?

Reply:
Then at least change the function to accept an `int` instead of `miopenConvolutionFindMode_t`. This way I only need to define the function in migraphx.

Reply:
Let me add another simple wrapper function that accepts int for you.

Reply:
BTW alternatively, I can move this stuff back t othe public API but hide it within
```C++
#ifdef MIOPEN_ENABLE_FIND_MODE_API
...
#endif
```
So only users that write:
```C++
#define MIOPEN_ENABLE_FIND_MODE_API
#include <miopen.h>
```
will be able to use this functionality.

Please let me know if you prefer this way.

Reply:
> BTW alternatively, I can move this stuff back to the public API but hide it within

No, I dont think its good to have this in the header.

Reply:
Ok

Reply:
In terms of a public API, I recommend that it should be more abstract, and not specific to convolutions. The abstraction will also provide flexibility moving forwards, and should have arguments that are conceptually clear. 
Example:
SetAlgoSearchLevel(  some uchar typedef with values right now: exhaustiveSearch=1,  ... , minimizeCompilation = 5 ) .  the params roughly correspond to the current find modes. I would avoid enums, and favor typedefs to uchar for this kind of thing, for forward and backward compatibility at the API level at least.

This API should be public, as different clients to miOpen have different requirements, and would want to control this programmatically based on the particular application (network+dataset).

Reply:
@hgaspar Good idea, thanks. However, that would be overdesign right now. We do not plan to add Find API for other primitives and do not foresee the need for it. Current Find Modes are only applicable for convolutions and can only be explained (documented) in convolution-specific terms.